src/audio.c

```diff
@@ -46,83 +147,127 @@ static int paudio_callback(const void *input_buffer,
-		if(info->channels == 1) {
-			for(i = even ? 0 : 1; i < frame_count; i += 2) {
-				pa_buffers[wp++] = input_samples[i];
-				if (wp >= PA_BUFF_SIZE) wp -= PA_BUFF_SIZE;
-			}
-		} else {
-			for(i = even ? 0 : 2; i < frame_count*2; i += 4) {
-				pa_buffers[wp++] = input_samples[i] + input_samples[i+1];
-				if (wp >= PA_BUFF_SIZE) wp -= PA_BUFF_SIZE;
-			}
-		}
-		/* Keep track if we have processed an even number of frames, so
-		 * we know if we should drop the 1st or 2nd frame next callback. */
-		if(frame_count % 2) even = !even;
+
+	const unsigned len = MIN(frame_count, PA_BUFF_SIZE - wp);
+	if(info->channels == 1) {
+		memcpy(pa_buffers + wp, input_samples, len * sizeof(*pa_buffers));
+		if(len < frame_count)
+			memcpy(pa_buffers, input_samples + len, (frame_count - len) * sizeof(*pa_buffers));
 	} else {
-		const unsigned len = MIN(frame_count, PA_BUFF_SIZE - wp);
-		if(info->channels == 1) {
-			memcpy(pa_buffers + wp, input_samples, len * sizeof(*pa_buffers));
-			if(len < frame_count)
-				memcpy(pa_buffers, input_samples + len, (frame_count - len) * sizeof(*pa_buffers));
-		} else {
-			for(i = 0; i < len; i++)
-				pa_buffers[wp + i] = input_samples[2u*i] + input_samples[2u*i + 1u];
-			if(len < frame_count)
-				for(i = len; i < frame_count; i++)
-					pa_buffers[i - len] = input_samples[2u*i] + input_samples[2u*i + 1u];
-		}
-		wp = (wp + frame_count) % PA_BUFF_SIZE;
+		for(i = 0; i < len; i++)
+			pa_buffers[wp + i] = input_samples[2u*i] + input_samples[2u*i + 1u];
+		if(len < frame_count)
+			for(i = len; i < frame_count; i++)
+				pa_buffers[i - len] = input_samples[2u*i] + input_samples[2u*i + 1u];
 	}
+	wp = (wp + frame_count) % PA_BUFF_SIZE;
+
 	pthread_mutex_lock(&audio_mutex);
 	write_pointer = wp;
 	timestamp += frame_count;
 	pthread_mutex_unlock(&audio_mutex);
 	return 0;
 }
 
-int start_portaudio(int *nominal_sample_rate, double *real_sample_rate)
+
+void resetBuffers(){
+	pthread_mutex_lock(&audio_mutex);
+		memset(pa_buffers, 0, sizeof(pa_buffers));
+		write_pointer = 0;
+
+		timestamp = 0;
+
+	pthread_mutex_unlock(&audio_mutex);
+}
+
+int start_portaudio(int *nominal_sample_rate, double *real_sample_rate,
+					char* hostdeviceName, char*hostAPIname,
+					gboolean bShowError)
 {
-	PaStream *stream;
+
+    debug("start_portaudio %d   %s   %s \n", *nominal_sample_rate, hostAPIname, hostdeviceName);
+
 
 	if(pthread_mutex_init(&audio_mutex,NULL)) {
 		error("Failed to setup audio mutex");
 		return 1;
 	}
 
+
+
 	PaError err = Pa_Initialize();
 	if(err!=paNoError)
 		goto error;
 
-#ifdef DEBUG
-	if(testing) {
-		*nominal_sample_rate = PA_SAMPLE_RATE;
-		*real_sample_rate = PA_SAMPLE_RATE;
-		goto end;
-	}
-#endif
+	resetBuffers();
 
-	PaDeviceIndex default_input = Pa_GetDefaultInputDevice();
-	if(default_input == paNoDevice) {
-		error("No default audio input device found");
-		return 1;
+
+	PaStreamParameters inputParameters;
+	inputParameters.channelCount = 2; // Stereo
```

Comment:
What's the point of setting these here?  And why stereo?  The audio is converted to mono so wouldn't mono be better to capture?